File: typescript/scheme/scheme_monochrome.ts
/**
 * @license
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {Hct} from '../hct/hct.js';
import {TonalPalette} from '../palettes/tonal_palette.js';

import {DynamicScheme} from './dynamic_scheme.js';
import {Variant} from './variant.js';

/** A Dynamic Color theme that is grayscale. */
export class SchemeMonochrome extends DynamicScheme {
  constructor(sourceColorHct: Hct, isDark: boolean, contrastLevel: number) {
    super({
      sourceColorArgb: sourceColorHct.toInt(),
      variant: Variant.MONOCHROME,
      contrastLevel,
      isDark,
      primaryPalette: TonalPalette.fromHueAndChroma(sourceColorHct.hue, 0.0),
      secondaryPalette: TonalPalette.fromHueAndChroma(sourceColorHct.hue, 0.0),
      tertiaryPalette: TonalPalette.fromHueAndChroma(sourceColorHct.hue, 0.0),
      neutralPalette: TonalPalette.fromHueAndChroma(sourceColorHct.hue, 0.0),
      neutralVariantPalette:
          TonalPalette.fromHueAndChroma(sourceColorHct.hue, 0.0),
    });
  }
}

File: cpp/scheme/scheme_monochrome.cc
/*
 * Copyright 2023 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/scheme/scheme_monochrome.h"

#include "cpp/cam/hct.h"
#include "cpp/dynamiccolor/dynamic_scheme.h"
#include "cpp/dynamiccolor/variant.h"
#include "cpp/palettes/tones.h"

namespace material_color_utilities {

SchemeMonochrome::SchemeMonochrome(Hct set_source_color_hct, bool set_is_dark,
                                   double set_contrast_level)
    : DynamicScheme(
          /*source_color_argb:*/ set_source_color_hct.ToInt(),
          /*variant:*/ Variant::kMonochrome,
          /*contrast_level:*/ set_contrast_level,
          /*is_dark:*/ set_is_dark,
          /*primary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 0.0),
          /*secondary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 0.0),
          /*tertiary_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 0.0),
          /*neutral_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 0.0),
          /*neutral_variant_palette:*/
          TonalPalette(set_source_color_hct.get_hue(), 0.0)) {}

SchemeMonochrome::SchemeMonochrome(Hct set_source_color_hct, bool set_is_dark)
    : SchemeMonochrome::SchemeMonochrome(set_source_color_hct, set_is_dark,
                                         0.0) {}

}  // namespace material_color_utilities
